### PR TITLE
S2U-29 5.1.3.3 Private Messages: Request a read receipt feature

### PR DESCRIPTION
**Pull request by JuanDavid102**
This is paired with https://github.com/sakaiproject/sakai-reference/pull/205

### CODE REVIEW

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java [hunk 1]
```diff
@@ -145,6 +148,7 @@ public class PrivateMessageManagerImpl extends HibernateDaoSupport implements Pr
   @Setter private PreferencesService preferencesService;
   @Setter private ServerConfigurationService serverConfigurationService;
   @Setter private FormattedText formattedText;
+  private NotificationService notificationService;
```
Comment (jesusmmp) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java:151]: Is this used?

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java [hunk 2]
```diff
@@ -1563,6 +1569,12 @@ public void markMessageAsReadForUser(final PrivateMessage message, final String
     	if (((PrivateMessageRecipientImpl) pvtMessage.getRecipients().get(i)).getUserId().equals(searchRecipient.getUserId())){
     		recordIndex = i;
     		if (! ((PrivateMessageRecipientImpl) recipientList.get(recordIndex)).getRead()) {
+          if (((PrivateMessageRecipientImpl) recipientList.get(recordIndex)).getReadReceipt() != null && ((PrivateMessageRecipientImpl) recipientList.get(recordIndex)).getReadReceipt()) {
```
Comment (jesusmmp) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java:1572]: We should respect spacing as better as possible...

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java [hunk 3]
```diff
@@ -1179,6 +1183,7 @@ else if (pf == null && oldPf != null && (oldPf.getAutoForward()==PrivateForumImp
 		  PrivateMessageRecipientImpl receiver = new PrivateMessageRecipientImpl(
 				  userId, typeManager.getReceivedPrivateMessageType(), contextId,
 				  isRecipientCurrentUser, bcc);
+      receiver.setReadReceipt(readReceipt);
```
Comment (jesusmmp) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java:1186]: Spacing

File: msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/PrivateMessagesTool.java [hunk 4]
```diff
@@ -2979,7 +2986,8 @@ private PrivateMessage processPvtMsgReplyAllSendHelper(boolean preview, Boolean
 		  }
 	  }
 	  if(!preview){
-	          prtMsgManager.sendPrivateMessage(rrepMsg, returnSet, isSendEmail());
+      prtMsgManager.sendPrivateMessage(rrepMsg, returnSet, isSendEmail(), booleanReadReceipt);
```
Comment (jesusmmp) [inline msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/PrivateMessagesTool.java:2989]: Spacing

File: msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages_eu.properties [hunk 5]
```diff
@@ -907,3 +907,10 @@ cdfm_faq_forum_title=FAQ Forum
 cdfm_faq_forum_description=Forum for frequently asked questions
 cdfm_faq_topic_title=FAQ Topic
 cdfm_faq_topic_description=Topic for frequently asked questions
+
+#S2U-29
```
Comment (mpellicer) [inline msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages_eu.properties:911]: @pablosanroman would you mind contributing these strings? Thanks!

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java [hunk 6]
```diff
@@ -1574,6 +1585,40 @@ public void markMessageAsReadForUser(final PrivateMessage message, final String
...
+    additionalHeaders.add("Subject: " + rb.getFormattedMessage("pvt_read_receipt_email_subject", pvtMessage.getTitle()));
+
+    User user = null;
+    User messageCreator = null;
+    try {
+      user = userDirectoryService.getUser(pvtRecipient.getUserId());
+      messageCreator = userDirectoryService.getUser(pvtMessage.getCreatedBy());
+    } catch (UserNotDefinedException e) { }
+
+    DateFormat df = new SimpleDateFormat("MM/dd/yyyy HH:mm:ss");
```
Comment (mpellicer) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java:1616]: This is hardcoded, do not hardcode date formats.

Comment (mpellicer) [reply msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java:1616]: Look at this
https://github.com/sakaiproject/sakai/blob/fe6ac1e2b32f57e4503efdbaf2ee467bb8b8a184/msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionTopicBean.java#L1195

File: msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages_eu.properties (hunk 5 above)
Comment (PabloSanRoman) [reply msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages_eu.properties:911]: Hi @mpellicer , here are the lines

cdfm_faq_forum_description=Galdera ohikoentzako foroa
cdfm_faq_topic_title=FAQ gaia
cdfm_faq_topic_description=Galdera ohikoentzako gaia

Comment (mpellicer) [reply msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages_eu.properties:911]: Thank you Pablo, strings updated.

I'm missing this one in EU

pvt_publish_to_faq=Publish to FAQ

Comment (PabloSanRoman) [reply msgcntr/messageforums-api/src/bundle/org/sakaiproject/api/app/messagecenter/bundle/Messages_eu.properties:911]: Ok @mpellicer !
pvt_publish_to_faq=Publikatu FAQ gisa

**Pull request merged**